nucypher/characters/lawful.py

```diff
@@ -720,23 +720,77 @@ def work_orders_for_capsules(self,
+            for capsule in work_order.tasks:
+                work_order_is_useful = False
+                if len(capsule) >= m:
+                    remaining_capsules.discard(capsule)
+                else:
+                    work_order_is_useful = True
+                    break
+
+            # If all the capsules are now activated, we can stop here.
+            if not remaining_capsules:
+                break
+
+            if not work_order_is_useful:
+                # None of the Capsules for this particular WorkOrder need to be activated.  Move on to the next one.
+                continue
+
+            remaining_work_orders.append(work_order)
+
+        return remaining_work_orders, remaining_capsules
+
+    def _reencrypt(self, work_order, retain_cfrags=False):
+
+        if work_order.completed:
+            raise TypeError(
+                "This WorkOrder is already complete; if you want Ursula to perform additional service, make a new WorkOrder.")
+
+        # We don't have enough CFrags yet.  Let's get another one from a WorkOrder.
+        try:
+            cfrags_and_signatures = self.network_middleware.reencrypt(work_order)
+        except NodeSeemsToBeDown as e:
+            # TODO: What to do here?  Ursula isn't supposed to be down.  NRN
+            self.log.info(f"Ursula ({work_order.ursula}) seems to be down while trying to complete WorkOrder: {work_order}")
+            return False, [] # TODO: return a grievance?
+        except self.network_middleware.NotFound:
+            # This Ursula claims not to have a matching KFrag.  Maybe this has been revoked?
+            # TODO: What's the thing to do here?  Do we want to track these Ursulas in some way in case they're lying?  567
+            self.log.warn(f"Ursula ({work_order.ursula}) claims not to have the KFrag to complete WorkOrder: {work_order}.  Has accessed been revoked?")
+            return False, [] # TODO: return a grievance?
+        except self.network_middleware.UnexpectedResponse:
+            raise # TODO: Handle this
+
+        cfrags = work_order.complete(cfrags_and_signatures)
+
+        # TODO: hopefully GIL will allow this to execute concurrently...
+        # or we'll have to modify tests that rely on it
+        self._completed_work_orders.save_work_order(work_order, as_replete=retain_cfrags)
+
+        the_airing_of_grievances = []
+        for capsule, pre_task in work_order.tasks.items():
+            if not pre_task.cfrag.verify_correctness(capsule):
+                # TODO: WARNING - This block is untested.
+                from nucypher.policy.collections import IndisputableEvidence
+                evidence = IndisputableEvidence(task=pre_task, work_order=work_order)
+                # I got a lot of problems with you people ...
+                the_airing_of_grievances.append(evidence)
+
+        if the_airing_of_grievances:
+            return False, the_airing_of_grievances
+        else:
+            return True, cfrags
```

Comment:
What do you think about adding a type hint to this method for `Tuple[bool, list]`?

Reply:
I actually think that throwing an exception would be better, but this function was intended to be executed in parallel, and it all depends on whether the parallel runner can properly track exceptions. Added the signature for now.